### DISCUSSION TITLE
What is the fastest way to serialize vector of some structs into js array of objects?

### DISCUSSION THREAD

**Discussion by romeerez** [napi-rs/napi-rs#548] [Q&A]
I'm writing a library for database queries, using tokio-postgres, I'm newbie at rust and just began and it already performs a bit faster than a node library! I'm so happy.

By benchmarking it seems that half time rust spends on serializing to JsObject, here is what I have:

```rust
let mut array = env.create_array_with_length(rows.len())?;

for (i, row) in rows.iter().enumerate() {
  let mut obj = env.create_object()?;
  obj.set_property(env.create_string("id")?, env.create_int32(row.get(0))?)?;
  obj.set_property(env.create_string("first_name")?, env.create_string(row.get(1))?)?;
  // setting more properties...
  array.set_element(i as u32, obj)?;
}

Ok(array)
```

What is the fastest way construct objects?
All objects in array have the same set of properties, can I use this knowledge somehow?

**Comment by Schniz** [comment]
I wonder what will be the performance difference between using `serde::Serialize`, which would be less explicit:

```rust
#[derive(serde::Serialize)]
struct NodeObj {
  id: String,
  first_name: String
  // ...
}

// then in code

env.to_js_value(&vec![NodeObj { ... }]);
```

**Comment by Brooooooklyn** [comment]
According to the benchmark result, I think the fastest way is serialize strut into string, and `env.create_string` to make it a `JsString`, then using the `JSON.parse` to decode it from the JavaScript side https://github.com/napi-rs/napi-rs/blob/main/bench/src/create_array.rs

 


```
createArrayJson:
    30 763 ops/s, ±1.09%   | fastest

  create array for loop:
    6 344 ops/s, ±1.05%    | slowest, 79.38% slower

  create array with serde trait:
    6 543 ops/s, ±1.06%    | 78.73% slower
```
